### ISSUE TITLE
Voxel map storage order and its implication.

### ISSUE THREAD

**Issue by versatran01** [KumarRobotics/kr_autonomous_flight#35]
https://github.com/KumarRobotics/autonomy_stack/blob/3231c89b163faa78fd1b018cf361514129a2601a/map_plan/pkgs/mapper/src/voxel_mapper.cpp#L156

The map storage is a boost multi_array which defaults to c_storage_order https://www.boost.org/doc/libs/1_63_0/libs/multi_array/doc/user.html#sec_storage
which is "row-major".

The iteration pattern is 
```
for (n(0) = 0; n(0) < dim_(0); n(0)++) {
    for (n(1) = 0; n(1) < dim_(1); n(1)++) {
      for (n(2) = 0; n(2) < dim_(2); n(2)++) {
	    int idx = n(0) + dim_(0) * n(1) + dim_(0) * dim_(1) * n(2);
        data[idx] = array[i][j][k]
```
Here `data` is just a vector which is 1d.
The computation of `idx` is col-major order. 
https://en.wikipedia.org/wiki/Row-_and_column-major_order#Address_calculation_in_general

Correct me if I'm wrong but this seems suboptimal.

**Comment by orlando21lara** [comment]
Hi, just to be sure that I understand the task correctly; it is desired to have the map have a storage order of column major instead of row major to avoid the conversion? In other words we want to be able to linearly copy the values from `map_` to `voxel_map.data` without the need of computing `idx` (i.e. voxel_map.data[k] = map_[k] with the appropriate clipping)? For convenience, I have copied and pasted the code below from the line mentioned in the first post.
```
  voxel_map.data.resize(dim_(0) * dim_(1) * dim_(2), val_default);
  Eigen::Vector3i n;
  for (n(0) = 0; n(0) < dim_(0); n(0)++) {
    for (n(1) = 0; n(1) < dim_(1); n(1)++) {
      for (n(2) = 0; n(2) < dim_(2); n(2)++) {
        if (map_[n(0)][n(1)][n(2)] > val_even) {
          int idx = n(0) + dim_(0) * n(1) + dim_(0) * dim_(1) * n(2);
          voxel_map.data[idx] = val_occ;
        } else if (map_[n(0)][n(1)][n(2)] >= val_free) {
          int idx = n(0) + dim_(0) * n(1) + dim_(0) * dim_(1) * n(2);
          voxel_map.data[idx] = val_free;
        }
      }
    }
  }
```

**Comment by fcladera** [comment]
@versatran01 

It would also be nice to benchmark the improvements in performance due to this change.

**Comment by versatran01** [comment]
There are a few places that use a voxel map. One is in the mapper class, the other is in jps/mpl. 
Currently, in the mapper class, the storage of this voxel map is a boost array with 3 dimensions, and by default boost array uses row-major storage order. On the other hand, the internal storage in jps/mpl is a simple vector of chars, which uses a column storage order.

One reason for using the same storage order is consistency. The other reason is efficiency, converting from row-major to column-major is not cache-friendly. Finally, if we use a column-major storage order, then it is easy to slice a 2d section of the map. Therefore, I think we should change the mapper to store data in column-major order. 

I suggest we just use std::vector instead of boost array for map storage. There are also a couple of other places where this data is copied unnecessarily, which can be changed to a move.

**Comment by orlando21lara** [comment]
You are suggesting a **one-dimensional** `std::vector` correct?

I only found instances of the VoxelMapper class (class that stores map in a boost array) in the local_global_mapper.cpp source file, but they do not directly use the map stored in the boost array since it is a private member. Also, the source files in jps and mpl use a voxel map but it is defined by the `planning_ros_msgs::VoxelMap` struct which stores the map in a one-dimensional `std::vector<int8>`.

With that said, should I only refactor the VoxelMapper class to use a one-dimensional `std::vector`, that is column-major stored, instead of a three-dimensional `boost:multi_array` that is currently row-major stored? I would also like to mention that I believe it is pretty straightforward to change the order storage in the boost array, if that is preffered.

**Comment by versatran01** [comment]
I suggest using the same type as in the VoxelMap which is std::vector<int8_t> https://github.com/KumarRobotics/kr_planning_msgs/blob/main/kr_planning_msgs/msg/VoxelMap.msg.
But a boost array with column-major is also possible. Although I don't really see why a boost::array is needed in the first place.
It also goes without saying that no PR will be merged without tests.
I recommend you do a first PR to just add tests (and maybe benchmarks as Fernando suggested) to the mapper and a second PR that does the actual refactoring.

**Comment by orlando21lara** [comment]
Ok thank you for the quick reply. I'll work on that then.

**Comment by versatran01** [comment]
Sure, thanks for helping out.

**Comment by orlando21lara** [comment]
I made the first PR #163, using googletest, but I didn't do any extensive testing. First I wanted to be sure if this is what is desired. As I mentioned in the PR, since we want to measure performance maybe I should use google/benchmark instead of googletest since googletest has a low precision to measure the time a test takes.

I also looked more into the code for the class and I believe that changing to an `std::vector` will require a refactoring of most of the methods in the class, not just `getMap` https://github.com/orlando21lara/kr_autonomous_flight/blob/6204563602738068f6e0c52bd1183cf361eccb89/autonomy_core/map_plan/mapper/src/voxel_mapper.cpp#L187

I would like to note that using an `std::vector` in `getMap` would remove the need to compute an indexing variable from row-major to column-major, but it would also **add** the need to compute an indexing variable in the other methods. These methods currently take 3D coordinates and use them to directly index the 3D boost array without the need of an indexing array. We would have do something similar to the voxel map used in jps:
https://github.com/orlando21lara/kr_autonomous_flight/blob/6204563602738068f6e0c52bd1183cf361eccb89/autonomy_core/map_plan/jps3d/src/map_util.cpp#L8
https://github.com/orlando21lara/kr_autonomous_flight/blob/6204563602738068f6e0c52bd1183cf361eccb89/autonomy_core/map_plan/jps3d/src/graph_search.cpp#L55

Of course, I can test both solutions to see what is more efficient, but I would like to know your thoughts.

Should I focus on writing an extensive set of tests for the Voxel Mapper class or should do benchmarking with google/benchmark?

**Comment by versatran01** [comment]
Yeah for microbenchmark you would want to use google benchmark. But we can do something even simpler. Just run both versions and put some timing stuff around the callback and eyeball it.

The tests should be unit tests, they should test specific functions of the mapper class to make sure that your changes don't break anything.

**Comment by orlando21lara** [comment]
Ok should I write unit test for all the functions that make use of the map variable or should I discard some?

Should I also consider functions that use the inflated_map variable since they are closely related. This variable also currently uses a boost array, so we might want to change that as well correct?

**Comment by versatran01** [comment]
You can talk to xu about what functions are needed and even remove the unused ones.

**Comment by orlando21lara** [comment]
I've made some more comprehensive tests on the methods that are currently being used and hopefully with this concludes the PR. I decided to make the tests to be friend classes of the VoxelMapper class to be able to test their private methods, let me know if this ok. Also after making the tests I found some bugs and pushed the corrections in my last commit. Any feedback is appreciated.

**Comment by orlando21lara** [comment]
I have finished gathering the performance results for the different data types used for the map. I used Google Benchmark for this purpose to run three benchmarks on four of the functions that are currently being tested with Google Test. Each benchmark  just changes the resolution (0.5, 0.25, 0.1) of the map which always has dimensions (x: 200, y: 200, z: 10). 

The results are shown in the two images below; the first one compares the perfomance of using row major stored boost multi array to a column major stored multi array while the second one compares it to using an std::vector.

**Row Major Boost Multi Array VS Column Major Boost Multi Array**
![BoostRow_VS_BoostCol](https://user-images.githubusercontent.com/33842080/173416099-9185cde0-d8e6-4d12-a8b8-14c015427200.png)

**Row Major Boost Multi Array VS std::vector**
![BoostRow_VS_Vector](https://user-images.githubusercontent.com/33842080/173416120-87584be5-b4d2-492e-9a78-14709826eb1b.png)

We can see that in both cases the getMap function is faster but for the other functions, we either don't get quicker times or they are approximately the same. 

I would like to note that the tests were not completely consistent; but in all the runs that I made, getMap was always faster and the other functions were always slower when using the std::vector. I believe this makes sense since we have to compute an index value to convert from 3D indices to 1D index. In the case of the column major stored boost array I changed the traversing order of the for loops to z -> y -> x to make it more cache friendly given the storage order; otherwise it would also always be slower.
